### ISSUE TITLE
Very slow when working with big textbox in Zepplin

### ISSUE THREAD

**Issue by GaryLiuTelus** [glacambre/firenvim#1047]
- OS Version: win x86-64
- Browser Version:  Firefox 88.0
- Browser Addon Version: 0.2.6
- Neovim Plugin Version: 0.2.6
- Neovim version 0.50
- What I tried to do: I tried to use this plugin with Zeppelin (similar to Jupyter Notebook), version 0.8
- What happened: It did well for smaller cells, but whenever it enters into a big cell with long code block (normally over screen height, i.e. some contents are outside screen), the speed is getting very very slow. Firefox will give the warning to stop the page, as it slow down whole Firefox session.

**Comment by glacambre** [comment]
This is surprising to hear, I'm getting 60 fps scrolling in fullscreen and that's with integrated graphics.

What is the size of the neovim frame in pixels? What CPU and GPU are you using?

I'm also going to need you to record a performance profile. To do this, start Firenvim in a big cell and fill the buffer with data (e.g. fill the first line with `ababab...` and the second line with `cdcdcd...`, copy them with `GGyG` and paste them 10000 times with `10000p`). Once the buffer is ready, visit https://profiler.firefox.com/ , enable the profiler, go back to the tab that contains Firenvim, reproduce the slowness for as long as possible and then stop the recording. This will open a new tab where you will find a button labeled "Upload Local Profile" in the top right corner. Click it, make sure all checkboxes are ticked, click "download" and upload the profile here.

**Comment by GaryLiuGTA** [comment]
[Firefox 2021-05-20 14.55 profile.json.gz](https://github.com/glacambre/firenvim/files/6518238/Firefox.2021-05-20.14.55.profile.json.gz)
Thanks for your quick response. Above is the profile. Please let me know if you need any further information. 

This is done using Jupyter Notebook running at local server (localhost). The buffer loading is quicker than Zeppelin, which is a remote server. And it is slow especially scroll up/down to show contents outside screen.

**Comment by glacambre** [comment]
>Please let me know if you need any further information.

I'd like to know what CPU & GPU you are using and the dimensions of the Firenvim frame in pixels :).

According to the performance profile you uploaded, the only slow thing is [putImageData](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/putImageData) - that's the one thing Firenvim has to use to draw text, so I can't optimize anything there. It being slow makes me think that there might be an issue with your Firefox/driver/OS installation that is slowing things down. Could you try [disabling (or enabling) hardware acceleration](https://support.mozilla.org/en-US/kb/hardware-acceleration-and-windowblinds-crash) and seeing if it changes anything?

**Comment by GaryLiuTelus** [comment]
CPU: Intel i7-6600U @2.6GHz
GPU Intel HD 520
Firenvim frame size is roughly 2490*1439(I am using 2K*2 monitor). even I worked on a smaller textbox, for every line I scroll beyond the current visible area, there is 0.5 to 1 second lag.
I tried disabled hardware acceleration, but it does not help. 

I can try this plugin in the other old laptop (with Ubuntu 20.04) during the weekend and see how it will go. 

Thanks for your information!

**Comment by GaryLiuGTA** [comment]
Hi, I tested in my old laptop, which has 3rd Gen i7 (i7 3250). And the firenvim ran very smoothly (I am replying this using firenvim on my old laptop now). So I guess there are something wrong with my new laptop.

Thanks!

**Comment by glacambre** [comment]
Thank you for trying another machine - can you confirm that the dimensions of the Firenvim frame were the same?

**Comment by GaryLiuGTA** [comment]
Yes, frame size is similar. And in the new laptop, the firenvim works fine in the regular small text box, such as those in Github, but every time after I use `set lines =  xxx` to change the size, it became a little bit of sluggish. But this did happen in my old laptop. Still no idea why it happened, but it could be due to some settings in Firefox or OS (Ubuntu vs. Windows 10).

**Comment by dclong** [comment]
In my cases, it takes about 1 minute to run `:set lines=30` in a JupyterLab notebook cell even if the cell contains just 1 line of text.

**Comment by ramfattah** [comment]
@glacambre, this happens in ChatGPT using Edge browser as well to the point where it becomes unuasble.

Here is the demo: at the end, firenvim completely froze:

[Screencast from 2024-11-07 23-44-17.webm](https://github.com/user-attachments/assets/424d8f14-2be6-4822-bc92-cad777ccf59e)


my firenvim config in lua:
```lua
-- Firenvim global and local settings
vim.g.firenvim_config = {
  globalSettings = {
    alt = 'all', -- Enable Alt key for all commands
  },
  localSettings = {
    ['.*'] = {
      cmdline = 'neovim',
      content = 'text',
      priority = 0,
      selector = 'textarea',
      takeover = 'never',
    },
  },
}

vim.api.nvim_create_autocmd({ 'BufEnter' }, {
  pattern = '*_*.txt',
  command = 'set filetype=text',
})

vim.api.nvim_create_autocmd({ 'TextChanged', 'TextChangedI' }, {
  nested = true,
  command = 'write',
})

-- Define Firenvim-specific UI settings
local function setup_firenvim()
  vim.opt.ruler = false -- Disable ruler
  vim.opt.showcmd = false -- Disable command display
  vim.opt.laststatus = 0 -- Hide the status line
  vim.opt.showtabline = 0 -- Hide the tab line
  vim.opt.number = true
  vim.opt.relativenumber = true
end

-- Autocommand group for Firenvim settings
vim.api.nvim_create_augroup('firenvim', { clear = true })
vim.api.nvim_create_autocmd('FileType', {
  pattern = 'text',
  callback = setup_firenvim,
  group = 'firenvim',
})

```

**Comment by glacambre** [comment]
@ramfattah Can't repro. Please use your browser to record a performance profile and upload it here.

**Comment by ramfattah** [comment]
@glacambre thanks for the quick response.

Attached is my edge recorded profiler below, please let me know if you'd like more info:
[edge_profiler.zip](https://github.com/user-attachments/files/17685917/edge_profiler.zip)

**Comment by ramfattah** [comment]
@glacambre

Also I want to mention that I can only repro this laggy behavior in `Edge` browser. 

Specifically after adding this config:
```lua
vim.api.nvim_create_autocmd({ 'TextChanged', 'TextChangedI' }, {
  nested = true,
  command = 'write',
})
```


But in FireFox, I'm not able to repro.

**Comment by glacambre** [comment]
@ramfattah Thanks. As the profile shows, edge gets stuck on a call to [getImageData](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getImageData):
![tmp](https://github.com/user-attachments/assets/fff32c1c-8380-4820-9ed2-b0068c2da3f0)

getImageData is a browser primitive, so this really sounds like a bug in edge. I would recommend attempting to [disable (or enable) hardware acceleration](https://appuals.com/disable-or-enable-hardware-acceleration/).

**Comment by ramfattah** [comment]
@glacambre,

Thank you very much. Disabling gpu in Edge fixed the issue.

Appreciate your prompt help :)

**Comment by glacambre** [comment]
This should be fixed in Fiirenvim 0.2.17.
